src/ex2/CompileVisitor.java

```diff
@@ -196,19 +197,135 @@ public void visit(WhileStatement whileStatement) {
+
+        var label3 = lastLabelNumber++;
+        var label4 = lastLabelNumber++;
+
+        addLine(String.format("br i1 %%_%s, label %%arr_check%s, label %%arr_check%s", indexComparisonToLengthRegister, label3, label4));
+        addLine(String.format("arr_check%s:", label3));
+        addLine("call void @throw_oob()");
+        addLine(String.format("br label arr_check%s", label4));
+        addLine(String.format("arr_check%s:",label4));
+        var indexPlusOneRegister = lastRegisterNumber++;
+        addLine(String.format("%%_%s = add i32 %%_%s, 1", indexPlusOneRegister, indexRegister));
+
+        var elementPointerRegister = lastRegisterNumber++;
+        addLine(String.format("%%_%s = getelementptr i32, i32* %%_%s, i32 %%_%s", elementPointerRegister, arrayRegister, indexPlusOneRegister));
+        var elementValueRegister = lastRegisterNumber++;
+        addLine(String.format("%%_%s = load i32, i32* %%_%s", elementValueRegister, elementPointerRegister));
     }
 
     @Override
     public void visit(ArrayLengthExpr e) {
-
+        e.arrayExpr().accept(this);
+        var arrayRegister = lastRegisterNumber - 1;
+        var lengthElementPointerRegister = lastRegisterNumber++;
+        addLine(String.format("%%_%s = getelementptr i32, i32* %%_%s, i32 0", lengthElementPointerRegister, arrayRegister));
+        var lengthRegister = lastRegisterNumber++;
+        addLine(String.format("%%_%s = load i32, i32* %%_%s", lengthRegister, lengthElementPointerRegister));
     }
 
     @Override
     public void visit(NewIntArrayExpr e) {
+        e.lengthExpr().accept(this);
+        var lengthResultRegister = lastRegisterNumber - 1;
+        var compareToZeroResultRegister = lastRegisterNumber++;
+
+        addLine(String.format("%%_%s = icmp slt i32 %%_%s , 0", compareToZeroResultRegister, lengthResultRegister));
+
+        var label1 = lastLabelNumber++;
+        var label2 = lastLabelNumber++;
+
+
+        addLine(String.format("br i1 %%_%s, label %%arr_alloc%s, label %%arr_alloc%s", compareToZeroResultRegister, label1, label2));
+        addLine("arr_alloc" + label1 + ":");
+        addLine("call void @throw_oob()\n");
+        addLine("br label %alloc_arr" + label2);
+        addLine("alloc_arr" + label2 + ":" );
+
+        var lengthPlusOneResultRegister = lastRegisterNumber++;
+
+        addLine(String.format("%%_%s = addi32 %%_%s, 1", lengthPlusOneResultRegister, lengthResultRegister));
+
+        var allocationRegister = lastRegisterNumber++;
+
+        addLine(String.format("%%_%s = call i8* @calloc(i32 4, i32 %%_%s)", allocationRegister, lengthPlusOneResultRegister));
+
+        var castRegister = lastRegisterNumber++;
+
+        addLine(String.format("%%_%s = bitcast i8* %%_%s to i32*", castRegister, allocationRegister));
+
+        addLine(String.format("store i32 %%_%s, i32* %%_%s", lengthResultRegister, allocationRegister));
 
```

Comment:
last line should have a pointer to the array so it can be accessed after creation

Reply:
The last line doesn't allocate a new register, so the last register (lastRegisterNumber-1) will be a register the register that holds the array